content/blog/2019-10-23-global-value-numbering/index.md

```diff
@@ -0,0 +1,357 @@
+  x : int = const 2;
+  jmp exit;
+exit:
+  print x;
+```
+
+To convert these basic blocks to Bril code, we rename variables and insert a `phi` instruction when the control flow merges again in the last block.
+
+```
+start:
+  br cond1 left right;
+left:
+  x1 : int = const 1;
+  jmp exit;
+right:
+  x2 : int = const 2;
+  jmp exit;
+exit:
+  x3 : int = phi x1 x2;
+  print x3;
+```
+
+Having unique variable assignments is enormously helpful in implementing optimization---most industrial-strength compilers, including [LLVM][] and [GCC][], rely on it internally. We followed a standard algorithm to convert Bril programs to SSA form. 
+
+The first pass inserts `phi` nodes wherever control flow merges multiple variable definitions. 
+We determine this condition by computing the [_dominance frontier_][domf] (essentially, boundaries in the control flow graph where control flow can recombine) for every basic block. 
+For every block in the dominance frontier of every variable, we insert a `phi` node with as many arguments as there are predecessors to the block. 
+One bug we hit with this part of the implementation was realizing that we only should insert `phi` nodes at the dominance frontier for variables that we assigned to multiple times. 
+For convenience of later analysis, we also augmented the base algorithm to track which basic blocks (or source) each `phi` argument originated from (for the above example, our generated `phi` would be `x3 : int = phi x1 x2 left right;`)  
+
+The second pass is a recursive renaming algorithm that uses a stack to determine the correct version of a variable name for every point in the program.
+The trickiest part of implementing this algorithm was traversing the basic blocks in the correct order.
+The high level algorithm specifies that you should recursively traverse the children of each block in the dominator tree. 
+However, we found that not only did we ned to traverse the children, but we needed to traverse them _in order_ relative to the control flow graph. 
+In particular, Bril’s ecosystem already had a utility for getting the dominators of every block. To get the dominance tree, we calculated the transitive reduction of this relation. 
+Our implementation additionally extending the dominance tree such that every list of children was ordered relative to a reverse post-order traversal of the control flow graph.
+
+In later testing (using an example, `test/gvn/constant-folding-tau-example.bril`, ported from the [class notes of Prof. Mooly Sagiv of Tel Aviv University)][tau], we found one minor bug with our SSA implementation. 
+Bril programs do not have a type checker and are aggressively dynamic: it’s perfectly valid to assign to a variable in one branch of a conditional but not the other. 
+The base SSA algorithm does not account for this, so we fail to properly rename in the second step of our algorithm. 
+For now, we work around this case by assigning a default value to to the variable before we branch. 
```

Comment:
Yeah, that's totally the right thing to do—to imagine that every variable is assigned to with a special "uninitialized" value at the entry block. I should probably mention this in future SSA lectures.